src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,211 @@
+void ReportSchedulerImpl::OnReadHandlerRemoved(ReadHandler * aReadHandler)
+{
+    UnregisterReadHandler(aReadHandler);
+}
+
+CHIP_ERROR ReportSchedulerImpl::RegisterReadHandler(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * newNode = FindReadHandlerNode(aReadHandler);
+    // If the handler is already registered, no need to register it again
+    VerifyOrReturnValue(nullptr == newNode, CHIP_NO_ERROR);
+    newNode = mNodesPool.CreateObject(aReadHandler, mTimerDelegate, ReportTimerCallback);
+    mReadHandlerList.PushBack(newNode);
+
+    ChipLogProgress(DataManagement,
+                    "Registered a ReadHandler that will schedule a report between system Timestamp: %" PRIu64
+                    " and system Timestamp %" PRIu64 ".",
+                    newNode->GetMinTimestamp().count(), newNode->GetMaxTimestamp().count());
+
+    Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    Milliseconds32 newTimeout;
+    // If the handler is reportable, schedule a report for the min interval, otherwise schedule a report for the max interval
+    if ((newNode->IsReportableNow()))
+    {
+        // If the handler is reportable now, just schedule a report immediately
+        newTimeout = Milliseconds32(0);
+    }
+    else if (IsReadHandlerReportable(aReadHandler) && (newNode->GetMinTimestamp() > now))
+    {
+        // If the handler is reportable now, but the min interval is not elapsed, schedule a report for the moment the min interval
+        // has elapsed
+        newTimeout = Milliseconds32(newNode->GetMinTimestamp().count() - now.count());
+
+        ReturnErrorOnFailure(ScheduleReport(newTimeout, aReadHandler));
+    }
+    else
+    {
+        // If the handler is not reportable now, schedule a report for the max interval
+        newTimeout = Milliseconds32(newNode->GetMaxTimestamp().count() - now.count());
+    }
+
+    ReturnErrorOnFailure(ScheduleReport(newTimeout, aReadHandler));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReportSchedulerImpl::ScheduleReport(Timeout timeout, ReadHandler * aReadHandler)
+{
+    // Verify the handler is still registered
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnError(nullptr != node, CHIP_ERROR_NOT_FOUND);
+
+    // Cancel Report if it is currently scheduled
+    CancelTimerForHandler(node);
+
+    if (!IsChunkedReport(aReadHandler))
+    {
+        StartTimerForHandler(node, timeout);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void ReportSchedulerImpl::CancelReport(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+    CancelTimerForHandler(node);
+}
+
+void ReportSchedulerImpl::UnregisterReadHandler(ReadHandler * aReadHandler)
+{
+    // Verify list is populated and handler is not null
+    VerifyOrReturn((!mReadHandlerList.Empty() || (nullptr == aReadHandler)));
```

Comment:
This comment does not match the code.

Why do we care about checking mReadHandlerList anyway?  Worst case, we will not find the node, right?  But it seems like even that would involve some serious programmer error somewhere, since ReadHandlers register with us in their constructor and stay registered until their destructor.